core/src/main/java/novoda/lib/sqliteprovider/util/DatabaseStructure.java

```diff
@@ -0,0 +1,226 @@
+package novoda.lib.sqliteprovider.util;
+
+import android.database.Cursor;
+import android.database.sqlite.SQLiteDatabase;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.TreeMap;
+
+import novoda.lib.sqliteprovider.sqlite.IDatabaseMetaInfo.SQLiteType;
+
+public class DatabaseStructure {
+
+    private static final String SELECT_TABLES_NAME = "SELECT name FROM sqlite_master WHERE type='table';";
+    private static final String PRAGMA_TABLE_INFO = "PRAGMA table_info('%1$s');";
+    private static final String PRGAMA_INDEX_LIST = "PRAGMA index_list('%1$s');";
+    private static final String PRGAMA_INDEX_INFO = "PRAGMA index_info('%1$s');";
+    private static final String COLUMN_PRIMARY_KEY_INDEX = "pk";
+    private static final String COLUMN_NAME = "name";
+    private static final String COLUMN_TYPE = "type";
+
+    private static final List<String> DEFAULT_TABLES = Collections.singletonList("android_metadata");
+
+    private final SQLiteDatabase database;
+
+    private List<String> createdTables;
+    private Map<String, List<String>> foreignTables = new TreeMap<>();
+    private Map<String, List<Constraint>> uniqueConstraints = new TreeMap<>();
+
+    public DatabaseStructure(SQLiteDatabase database) {
+        this.database = database;
+    }
+
+    public List<String> tables() {
+        final Cursor tablesCursor = database.rawQuery(SELECT_TABLES_NAME, null);
+        createdTables = new ArrayList<>(tablesCursor.getCount());
+        parseTables(tablesCursor);
+        tablesCursor.close();
+        return Collections.unmodifiableList(createdTables);
+    }
+
+    private void parseTables(Cursor tablesCursor) {
```

Comment:
Hmm, not quite, since `parseTables()` then calls `addCreatedTableIfNotDefault()`, which may or may not add a table name to the list, so I would have to use the list as a collector for the later method.